### PR TITLE
Parallel Pipeline stage for BPZ

### PR DESCRIPTION
**Pull request by joezuntz**
Hi Sam,

This PR adds a pipeline stage designed to interface with TXPipe in particular in a few ways.  It:

- runs in parallel, so we can make best use of NERSC
- takes the same photometry_catalog format that is used in TXPipe
- saves to the same format as the MLZ version that we have
- runs on the metacal variants of the magnitudes as well

I'd be happy to just keep this in TXPipe if you prefer - just let me know.

Best,
Joe

### CODE REVIEW

File: bpzpipe/BPZPipe2.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+            # Clip to min mag errors
+            np.clip(mag_errs, self.config['mag_err_min'], 1e10, mag_errs)
+
+            # Convert to pseudo-fluxes
+            flux = 10.0**(-0.4*mags)
+            flux_err = flux * (10.0**(0.4*mag_errs) - 1.0)
+
+            # Check if an object is seen in each band at all.
+            # Fluxes not seen at all are listed as infinity in the input,
```
Comment (sschmidt23) [inline bpzpipe/BPZPipe2.py:208]: Fluxes are not listed as infinity in the input, the convention that BPZ follows is that non-detections have a reported magnitude of 99.0, and in place of a magnitude error the mag_error column should contain the 1-sigma flux limit.  BPZ then replaces the flux for non-detections with zero, and the flux-error with the 1-sigma limit converted to flux

File: bpzpipe/BPZPipe2.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+            # so will come out as zero flux and zero flux_err.
+            # Check which is which here, to use with the ZP errors below
+            seen1 = (flux > 0) & (flux_err > 0)
```
Comment (sschmidt23) [inline bpzpipe/BPZPipe2.py:211]: As per the comment above, this definition of seen is not the usual convention

File: bpzpipe/BPZPipe2.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+            seen = np.where(seen1)
+            unseen = np.where(~seen1)
```
Comment (sschmidt23) [inline bpzpipe/BPZPipe2.py:213]: I would replace this with 
```
nondetect = 99.
nondetect_flux = 10.**(-0.4*nondetect)
unseen = np.isclose(flux,nondetect_flux,atol=nondetectflux*0.5)
flux[unseen] = 0.
flux_err[unseen] = 10.**(-0.4*np.abs(mag_errs[unseen]))
```
or something similar

File: bpzpipe/BPZPipe2.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+                flux  = data[f'flux{suffix}'][i]
+                flux_err = data[f'flux_err{suffix}'][i]
+
+                # and compute the PDF for it
+                pdf = self.estimate_pdf(flux_templates, kernel, flux, flux_err, mag_0, z)
+                
+                if suffix=="":
+                    pdfs[i] = pdf
+
+                # The pdf already sums to unity, so this gives us the mean
```
Comment (sschmidt23) [inline bpzpipe/BPZPipe2.py]: The default point estimate for BPZ is the maximum value of the marginalized posterior, not the mean.  If you want to use the mean in place of the mode that's fine, but for testing and comparison to previous BPZ runs I replaced the following line of code with 
`
point_estimates[s,i] = z[np.argmax(pdf)]
`

**Review by sschmidt23 — CHANGES_REQUESTED** [review]
Two small issues: the treatment of non-detections is different from how BPZ is usually run, and when I ran the original pipeline I got a vector of 'nan' for galaxies with mag = 99.0 in at least one band.  

Second, BPZ takes the mode of the posterior as the point estimate, not the mean, so if you want to compare with old runs for validation, or if you prefer the mode to the mean for a point redshift estimate, this should be changed (or we could store both).  If metacal is going to use a different convention from the mag = 99 mag_err = 1sigma, then you should still be sure that the flux_err is set to something like 1sigma detection threshold and is not somehow converted to a very small number, as that can lead to spurious results when computing chi^2/likelihoods.

Once those two changes were made, my simple example outputs had point estimates and p(z) posteriors that were exactly reproduced from the old/slow/messy BPZ, so everything else looks good

File: bpzpipe/BPZPipe2.py (hunk 4 above)
Comment (joezuntz) [reply bpzpipe/BPZPipe2.py]: I'll make this an option in the input.

File: bpzpipe/BPZPipe2.py (hunk 1 above)
Comment (joezuntz) [reply bpzpipe/BPZPipe2.py:208]: What I meant to say here is that the input photometry_catalog.hdf files that I'm mocking currently use infinity for undetected.  I'll add an option to make this work for both cases.

Although it looks like I'm currently setting mag_err=inf for undetected objects, which clearly makes no sense, so I'll need to deal with that as well.

File: bpzpipe/BPZPipe2.py (hunk 3 above)
Comment (joezuntz) [reply bpzpipe/BPZPipe2.py:213]: I'm a bit confused by this line, which I do find in the BPZ code as well:

```
flux_err[unseen] = 10.**(-0.4*np.abs(mag_errs[unseen]))
```
this seems to imply that larger mag errors give smaller flux errors?  Is the negative sign right?

File: bpzpipe/BPZPipe2.py (hunk 4 above)
Comment (joezuntz) [reply bpzpipe/BPZPipe2.py]: Done & pushed.

File: bpzpipe/BPZPipe2.py (hunk 3 above)
Comment (sschmidt23) [reply bpzpipe/BPZPipe2.py:213]: Yes, in that we assume that the mag_error column for "unseen" objects does not actually encode a magnitude error, it instead holds the 1 sigma limit for the observation (in units of magnitudes).  The thinking is that if the object is undetected, all you really know for sure is that the flux is somewhere below the noise from the background, and so it assigns an error equal to the 1 sigma background.  So, if the sky background limited you to, say, observing a 27.5 magnitude point source at the 1 sigma level, you would put 27.5 here, whereas if it was a deeper observation you would have e.g. 28.0.  Having a fainter 1 sigma limit should lead to smaller errors, so the minus sign is correct.

It's a confusing convention for sure, I think it was just a way to use the same column space without defining a separate column/variable for some objects that holds the background estimate.

**Pull request merged**